### PR TITLE
Merge #3509 TabletFile validation updates from 2.1 into main

### PR DESCRIPTION
**Pull request by cshannon**
Merges in changes from #3509 and refactors changes to be compatible with the previous refactoring of TabletFile

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/metadata/ValidationUtil.java [hunk 1]
```diff
@@ -53,6 +53,7 @@ public static Path validate(Path path) {
 
   public static void validateRFileName(String fileName) {
     Objects.requireNonNull(fileName);
+    // TODO: In 3.0.0 validate that filename starts with FilePrefix
```
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/metadata/ValidationUtil.java:56]: You should be able to implement this todo here and remove the comment. Something like `FilePrefix.fromPrefix(fileName.substring(0,1));`

**Review by dlmarion — COMMENTED** [review]
Looks ok to me. I think you can resolve the TODO

**Review by dlmarion — APPROVED** [review]

**Comment by cshannon** [comment]
I'll make sure all the checks are good and then I can merge into main

**Review by ctubbsii — CHANGES_REQUESTED** [review]
Changes are fine, but marking "request changes" instead of "approve" to get your attention on the following:

Please be very careful when merging this into main.

This contains merge commits as well as new commits. If this is merged from the GitHub UI using the squash method, it will drop the merge commit, and only keep the total changes. A subsequent merge commit (`-s ours`) will need to be done to include the commit history from 2.1. If it is merged  without that, then it will create an additional merge commit on top of the existing merge commit, and it won't squash the subsequent additional intermediate changes from the code review.

This is why we don't typically do merges as PRs, because GitHub's UI is inadequate for doing it correctly. I recommend merging this from the command-line instead, and just closing this PR after the review. Leave a comment in this PR with the merge commit hash.

**Comment by cshannon** [comment]
> Looks like there are test failures anyway. Not sure if those are merge artifacts or implementation issues. In any case, in future scenarios like this, I recommend just merging forward the other commit with `-s ours`, and then creating a new PR to reimplement the equivalent changes in the new branch that requires re-review. That way, it can just be squash-merged directly when it's ready.

Yeah it was from the extra validation commit. I actually realized when you were posting that merging this would break things with the extra merge commit. I mostly had just wanted Dave to take a look at it to make sure it was good.

In this case I did merge using -X ours and I ended up fixing things in the merge but next time I'll just do the fix in another PR after the merge.

**Comment by ctubbsii** [comment]
> In this case I did merge using -X ours and I ended up fixing things in the merge but next time I'll just do the fix in another PR after the merge.

`-X ours` is not the same as `-s ours`, and `-X ours` should be done very carefully, because it can still make changes to the destination branch. The `-X` version is a merge strategy **option** to the default merge strategy (which can vary from one git version to another). Typically, `-X` option says `merge using the default strategy, but select 'ours' when there's a conflict`. That's not usually what we want. The `-s` changes the entire merge strategy. Typically when we're merging this way, we want the destination branch we are merging *into* to be untouched by the merge... we want to connect the history only, but we want to keep the current code. That's when you should select the `-s ours` merge *strategy*, rather than the *option* on the default strategy. Usually, you should avoid `-X`, because otherwise, you'll have to try to figure out what it changed and what it kept... and it's not always obvious how to fix the conflicts it creates (and it's easy to overlook them). I find `-X` to result in far more merge bugs.

**Comment by cshannon** [comment]
> > In this case I did merge using -X ours and I ended up fixing things in the merge but next time I'll just do the fix in another PR after the merge.
> 
> `-X ours` is not the same as `-s ours`, and `-X ours` should be done very carefully, because it can still make changes to the destination branch. The `-X` version is a merge strategy **option** to the default merge strategy (which can vary from one git version to another). Typically, `-X` option says `merge using the default strategy, but select 'ours' when there's a conflict`. That's not usually what we want. The `-s` changes the entire merge strategy. Typically when we're merging this way, we want the destination branch we are merging _into_ to be untouched by the merge... we want to connect the history only, but we want to keep the current code. That's when you should select the `-s ours` merge _strategy_, rather than the _option_ on the default strategy. Usually, you should avoid `-X`, because otherwise, you'll have to try to figure out what it changed and what it kept... and it's not always obvious how to fix the conflicts it creates (and it's easy to overlook them). I find `-X` to result in far more merge bugs.

Makes sense, in this case I wanted a couple of changes except for the TabletFile changes so using -X brought that over but it was annoying going through and fixing everything. Next time I definitely plan to just use `-s ours `and just skip and re-implement the commit entirely as it would be easier. I'm used to using -X as a force of habit for merging forward things that just have version conflicts as it is handy for that.

One thing I thought of is if you use `git merge -s ours` does it only skip conflicting commits or will it not bring in any commits? Would it sill merge in commits from the source branch that merge cleanly?

**Comment by ctubbsii** [comment]
> I'm used to using -X as a force of habit for merging forward things that just have version conflicts as it is handy for that.

Oh yeah... great example. I should have used that after the 2.1.1 release. Instead, I just manually resolved the conflicts (using vim, one file at a time, from an SSH terminal on my phone, in the waiting room while my vehicle was getting an oil change and tire rotation). That would have **really** saved me the effort!!! :face_exhaling:

> One thing I thought of is if you use `git merge -s ours` does it only skip conflicting commits or will it not bring in any commits? Would it sill merge in commits from the source branch that merge cleanly?

It brings in references to any number of HEADs being merged in, but does not change anything in the destination branch. So, if there were more than one commit in the history that you're merging in this way, you should merge that forward the regular way first. I always check that the parent is merged first, before merging anything with `-sours`.

For example:

```bash
$ git remote update upstream
$ git checkout main && git merge upstream/main --ff-only # make sure working branch is up-to-date
$ git merge upstream/2.1~1 # clean merge of 2.1's HEAD's parent commit with changes, adds one merge commit to main branch
$ git merge upstream/2.1 -sours # include 2.1's HEAD without changes, adds second merge commit to main branch
$ doSomeEditing # reimplement 2.1's HEAD in main branch
$ git commit -a # adds a third commit to main branch for the reimplementation
$ git push upstream main
```

**Pull request merged**